[SPARK-6364] [MLlib] Implement equals and hashcode for Matrix

MechCoder opened the pull request:
Implementing methods equals hashcode and equals for Matrix

MechCoder committed:
[SPARK-6364] Implement equals and hashcode for Matrix

MechCoder commented:
cc @mengxr . Please review.

SparkQA commented:
[Test build #28781 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28781/consoleFull) for   PR 5081 at commit [`4041b59`](https://github.com/apache/spark/commit/4041b59aca38b18bb7a07d98621835a8d3543ba5).
 * This patch merges cleanly.

MechCoder commented:
@mengxr
1. Is there any reason why `forEachActive` is private? It took me some time to realize that it existed (and I already wrote my implementation by then)
2. The hashcode is just a generalization of that in the Vectors. I'm unable to understand why it is implemented like that. Would be great if you could give me some idea.

SparkQA commented:
[Test build #28783 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28783/consoleFull) for   PR 5081 at commit [`b5033b5`](https://github.com/apache/spark/commit/b5033b54bd4f7b6d5ec7be842c68be28cf59fdad).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28781 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28781/consoleFull) for   PR 5081 at commit [`4041b59`](https://github.com/apache/spark/commit/4041b59aca38b18bb7a07d98621835a8d3543ba5).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28781/
Test FAILed.

SparkQA commented:
[Test build #28785 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28785/consoleFull) for   PR 5081 at commit [`bf0e0e7`](https://github.com/apache/spark/commit/bf0e0e70a891a7355f2a5401e08803a32eaed333).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28783 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28783/consoleFull) for   PR 5081 at commit [`b5033b5`](https://github.com/apache/spark/commit/b5033b54bd4f7b6d5ec7be842c68be28cf59fdad).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28783/
Test FAILed.

MechCoder commented:
Also, I'm not sure what the failed tests are about.

SparkQA commented:
[Test build #28785 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28785/consoleFull) for   PR 5081 at commit [`bf0e0e7`](https://github.com/apache/spark/commit/bf0e0e70a891a7355f2a5401e08803a32eaed333).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28785/
Test FAILed.

srowen commented:
What's the use case for comparing matrices for equality? I think it's rarer than for vectors. Is it just for testing, mostly?

MechCoder commented:
Yes, I required it in this PR https://github.com/apache/spark/pull/5048/ to test that the matrices after serializing and deserializing are the same. (Also given the fact that it was already implemented for DenseMatrices, I just refactored it and added support for SparseMatrices)

mengxr commented:
The hashCode implementation was taken from Effective Java: http://stackoverflow.com/questions/10915309/need-explanation-for-hashcode-example-in-effective-java-textbook

`foreachActive` is a private API because we need to consider what operators to expose together.

@srowen It may help if we implement hashCode and content equals for both vectors and matrices. It is mostly used for testing.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala:
```diff
@@ -100,6 +100,50 @@ sealed trait Matrix extends Serializable {
   private[spark] def foreachActive(f: (Int, Int, Double) => Unit)
+
+  override def hashCode(): Int = {
+    var result: Int = numRows * numCols + 31
```
Hash `numRows` and `numCols` separately.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala:
```diff
@@ -100,6 +100,50 @@ sealed trait Matrix extends Serializable {
+    this.foreachActive { case (rowInd, colInd, value) =>
+      // ignore explict 0 for comparison between sparse and dense
+      if (value != 0) {
+        result = 31 * result + rowInd + (numRows * colInd)
```
Ditto. Hash row index and column index separately.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala:
```diff
@@ -100,6 +100,50 @@ sealed trait Matrix extends Serializable {
+        result = 31 * result + (bits ^ (bits >>> 32)).toInt
+      }
+    }
+    result
```
Hash `isTransposed`.

MechCoder replied:
Should this is be done? Even if `isTransposed` is not same, two matrices can be equal.

MechCoder replied:
Could you please verify this?
There might be cases where a CSR matrix is the same as a CSC matrix. See (https://github.com/apache/spark/pull/5081/files#diff-8fbb9a5e1adf997a37f4d05521b8a1acR488)
If I hash `isTransposed` I would give different hashes for both of these, which are not desirable.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala:
```diff
@@ -100,6 +100,50 @@ sealed trait Matrix extends Serializable {
+                case (i, j, value) => if (value != sm2(i, j)) return false
+              }
+            } else {
+                if (sm1.values != sm2.values) return false
```
This is not correct because one matrix may contain explicit zeros. Please include this case in the unit test.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala:
```diff
@@ -100,6 +100,50 @@ sealed trait Matrix extends Serializable {
+                if (sm1.rowIndices != sm2.rowIndices) return false
+            }
+            true
+          case (dm1: DenseMatrix, sm1: SparseMatrix) => Matrices.equals(dm1, sm1)
```
`sm1` -> `sm2`?

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala:
```diff
@@ -100,6 +100,50 @@ sealed trait Matrix extends Serializable {
+            if (sm1.isTransposed != sm2.isTransposed) {
+              if (sm1.values.length != sm2.values.length) return false
+              sm1.foreachActive {
+                case (i, j, value) => if (value != sm2(i, j)) return false
```
`sm2(i, j)` is quite expensive. Could you leave a TODO here to reduce the cost?

mengxr commented:
@srowen There are two approaches:

1. implement hashCode/equals
2. use the default implementation and use toArray to compare matrices in unit tests

Which do you prefer?

MechCoder commented:
I'm starting to think we should just test if `.toArray` is equal and this PR might be overkill if it is just for testing purposes.

mengxr commented:
Okay, let's keep this PR open and could you update the MatrixUDT PR?

MechCoder commented:
Done already :+1:

SparkQA commented:
[Test build #31050 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31050/consoleFull) for   PR 5081 at commit [`bf0e0e7`](https://github.com/apache/spark/commit/bf0e0e70a891a7355f2a5401e08803a32eaed333).
 * This patch **does not merge cleanly**.

andrewor14 commented:
@mengxr @MechCoder what is the status on this PR? It seems to fairly outdated at this point. Should we move forward or close it?

MechCoder commented:
Please do not. We just had a discussion about this yesterday.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35250 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35250/consoleFull) for   PR 5081 at commit [`e6b4cc5`](https://github.com/apache/spark/commit/e6b4cc5d977163a354e1906a3c7269af263c05d2).

SparkQA commented:
[Test build #35250 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35250/console) for   PR 5081 at commit [`e6b4cc5`](https://github.com/apache/spark/commit/e6b4cc5d977163a354e1906a3c7269af263c05d2).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

MechCoder commented:
@mengxr Please have a look at https://github.com/apache/spark/pull/6904/ This will help in ruling out some cases with minimum code repetition.

MechCoder committed:
improved hashing

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35299 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35299/consoleFull) for   PR 5081 at commit [`9d29561`](https://github.com/apache/spark/commit/9d295612355d0a0c1aa6a6944ae99e34f3fd806e).

SparkQA commented:
[Test build #35299 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35299/console) for   PR 5081 at commit [`9d29561`](https://github.com/apache/spark/commit/9d295612355d0a0c1aa6a6944ae99e34f3fd806e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35303 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35303/consoleFull) for   PR 5081 at commit [`7f46d17`](https://github.com/apache/spark/commit/7f46d17eb7511c0f3395fa8f99e7713b7ad5f7b3).

SparkQA commented:
[Test build #35303 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35303/console) for   PR 5081 at commit [`7f46d17`](https://github.com/apache/spark/commit/7f46d17eb7511c0f3395fa8f99e7713b7ad5f7b3).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

MechCoder commented:
There seems to be a unrelated whitespace error in L30 in `/spark/examples/src/main/scala/org/apache/spark/examples/DFSReadWriteTest.scala`

JoshRosen commented:
Yep, we're fixing it now via a hotfix.

andrewor14 commented:
should be fixed now. retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #35305 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35305/consoleFull) for   PR 5081 at commit [`7f46d17`](https://github.com/apache/spark/commit/7f46d17eb7511c0f3395fa8f99e7713b7ad5f7b3).

SparkQA commented:
[Test build #35305 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/35305/console) for   PR 5081 at commit [`7f46d17`](https://github.com/apache/spark/commit/7f46d17eb7511c0f3395fa8f99e7713b7ad5f7b3).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.
